feat: extend database logger context

lotyp opened the pull request:
## Changes

- Introduced additional options in DriverConfig for detailed logging control: `logInterpolatedQueries`. Converts query into interpolated one. (Before change, this was done using `logQueryParameters`)
- `logQueryParameters` - Now, does, what variable name says — Logs query parameters as array

- Modified statement method in Driver.php to utilize new logging options, enriching log context with driver details and query parameters.

## Rationale
These changes aim to improve the ORM's debugging and monitoring capabilities by offering detailed and configurable logging insights.

## Checklist

- Closes # -
- How was this tested:
    - Added new tests to cover context parameters and driver

    - Created example application with sample entity, fired persist to get logger output, which now looks like this:

	  - When `logInterpolatedQueries => true` and `logQueryParameters' => true`:
      ```
        [2024-03-19 21:42:14] local.INFO: INSERT INTO "posts" ("title", "content", "views", "id") VALUES ('title', 'content', 123, '018e58aa-38f9-725a-92cd-e69309ac5424') {"driver":"SQLite","elapsed":0.0006749629974365234,"rowCount":1,"parameters":["'title'","'content'","123","'018e58aa-38f9-725a-92cd-e69309ac5424'"]}
      ```

	- When `logInterpolatedQueries => false` and `logQueryParameters' => true`:
	  ```
        [2024-03-19 21:52:20] local.INFO: INSERT INTO "posts" ("title", "content", "views", "id") VALUES (?, ?, ?, ?) {"driver":"SQLite","elapsed":0.0006299018859863281,"rowCount":1,"parameters":["'title'","'content'","123","'018e58b3-7727-7185-9b1a-b8993069a367'"]}
      ```

    - Default behavior (both parameters set to `false`):
      ```
        [2024-03-19 21:54:17] local.INFO: INSERT INTO "posts" ("title", "content", "views", "id") VALUES (?, ?, ?, ?) {"driver":"SQLite","elapsed":0.0006580352783203125,"rowCount":1}
      ```

lotyp committed:
feat: extend database logger context

lotyp committed:
chore: remove unused function

lotyp committed:
chore: apply phpcs

lotyp committed:
tests: cover logger context parameters

lotyp committed:
tests: check if driver is set in logger context

lotyp committed:
chore: comply with style-ci

lotyp commented on src/Query/Interpolator.php:
```diff
@@ -141,4 +132,27 @@ private static function escapeStringValue(string $value): string
             '\\' => '\\\\',
         ]);
     }
```
Probably psalm typings are needed here

lotyp commented on src/Query/Interpolator.php:
```diff
@@ -141,4 +132,27 @@ private static function escapeStringValue(string $value): string
+     *
+     * @param iterable $parameters Parameters to categorize.
+     *
+     * @return array An associative array with keys 'named' and 'unnamed'.
```
Probably psalm typings are needed here, to define that array has two keys: 'named', 'unnamed'. Or rollback this change to initial.

msmakouz replied:
Yes, I agree with you. According to the code, there can only be two keys in the array, I would specify them in Psalm's types.

msmakouz commented on src/Driver/Driver.php:
```diff
@@ -676,16 +685,25 @@ protected function getPDO(): PDO|PDOInterface
+     * @return array
      */
-    protected function defineLoggerContext(float $queryStart, PDOStatement|PDOStatementInterface|null $statement): array
+    protected function defineLoggerContext(float $queryStart, PDOStatement|PDOStatementInterface|null $statement, ?iterable $parameters): array
```
I think we can use the `iterable` type for **$parameters** with a default value of an empty array. Since you pass parameters to this method with logging enabled (and they cannot be null, according to the typing of the `statement` method), or an empty array. Also, on line **703**, you use a foreach loop without checking **$parameters** for **null** before it: 
```suggestion
    protected function defineLoggerContext(
        float $queryStart, 
        PDOStatement|PDOStatementInterface|null $statement, 
        iterable $parameters = []
    ): array
```

lotyp replied:
I've updated the method to default $parameters to an empty array and removed the nullable type hint, making it always iterable. This change allows us to skip the null check before the foreach loop, streamlining the code as per your suggestion.

lotyp committed:
fix: parameters should always be iterable.

Default Value for `$parameters` — The `$parameters` argument now defaults to an empty array ([]). This ensures that it's always iterable, making the function more predictable and eliminating the need for null checks before iterating over $parameters.

lotyp committed:
chore: update CHANGELOG.md

roxblnfk reviewed (approved):
LGTM
